src/platform_linux/platform.h

```diff
@@ -734,51 +788,161 @@ platform_strtok_r(char *str, const char *delim, platform_strtok_ctx *ctx);
-         id, (p), STRINGIFY(p), __func__, __FILE__, __LINE__);                 \
-      (p) = NULL;                                                              \
+      debug_assert(((mf) != NULL),                                             \
+                   "Attempt to free a NULL ptr from '%s', line=%d",            \
+                   __func__,                                                   \
+                   __LINE__);                                                  \
+      platform_do_free((mf)->hid, (mf)->addr, (mf)->size, STRINGIFY(mf));      \
+      ZERO_STRUCT(*mf);                                                        \
    } while (0)
+/*
+ * ----------------------------------------------------------------------------
+ * void = platform_free_mem(platform_heap_id hid, void *p, size_t size);,
+ * void = platform_do_free(platform_heap_id hid, void *p, size_t size,
+ *                             const char *objname);
+ *
+ * Free a memory fragment at address 'p' of size 'size' bytes.
+ *
+ * These exist to facilitate re-cycling of free'd fragments in a shared-memory
+ * usage. That machinery works off of the fragment's 'size', hence we need to
+ * provide 'size' through this interface.
+ * ----------------------------------------------------------------------------
+ */
+// clang-format off
+#define platform_free_mem(hid, p, size)                                        \
+        platform_do_free((hid), (p), (size), STRINGIFY(p))
 
-#define platform_free_volatile(id, p)                                          \
+#define platform_do_free(hid, p, size, objname)                              \
+      platform_free_from_heap((hid), (p), (size),                              \
+                              (objname), __func__, __FILE__, __LINE__)
+
+// clang-format on
+
+/*
+ * ----------------------------------------------------------------------------
+ * void = platform_free_volatile(platform_heap_id hid,
+ *                               platform_memfrag *p)
+ *
+ * Similar to platform_free(), except it exists to free volatile ptr to
+ * allocated memory. The interface expects that the (single-) caller has
+ * packaged the memory fragment to-be-freed in a platform_memfrag *p * arg.
+ * There is just one consumer of this interface, so we don't go to the full
+ * distance as its sibling interface, to do error checking of args etc.
+ * ----------------------------------------------------------------------------
+ */
+#define platform_free_volatile(hid, p)                                         \
    do {                                                                        \
-      platform_free_volatile_from_heap(                                        \
-         id, (p), STRINGIFY(p), __func__, __FILE__, __LINE__);                 \
-      (p) = NULL;                                                              \
+      debug_assert(((p) != NULL),                                              \
```

Comment:
Once you change `platform_free_volatile_from_heap` to take a memfrag, you won't need all this type checking.